### DISCUSSION TITLE
How would you implement roles access with Adonis 5 ?

### DISCUSSION THREAD

**Discussion by yann-yinn** [adonisjs/core#3183] [Help]
Hi ! I'm working on a small adonis 5 monolithic starter ( https://github.com/yann-yinn/adonis-starter ) with ready-to-use authentication (login form, logout, and soon mails, reset password etc) to build some MVPs. I would like to add roles and permissions. Ex:

Role Redactor can create / read / delete articles.
Role Reviewer can only read article in draft mode.

I saw that there are now some bouncers, which works very well, but should we use them to implement a role system, and if yes, how would you link bouncers and roles in the code ?

**Comment by yann-yinn** [comment]
This is how I did for now. My users table has a "roles" column in database, which contains an array of roles ids. 

Roles are defined in code, not in database (this is easier do deploy)

**config/roles.ts**

```js
import { Roles } from "App/types";

const roles: Roles[] = [
  {
    id: "member",
    label: "Member",
  },
  {
    id: "admin",
    label: "Admin",
  },
];

export default roles;
```
**config/bouncer.ts**

```js
function userHasRoles(roles: string[], user: User) {
  let authorized = false;
  for (const role of roles) {
    if (user.roles.includes(role)) {
      authorized = true;
    }
  }
  return authorized;
}

export const { actions } = Bouncer.
  .define("createPost", (user: User) => {
    return userHasRoles(["admin", "member"], user);
  })
  .define("viewUsersList", (user: User) => {
    return userHasRoles(["admin"], user);
  })
  .define("createUser", (user: User) => {
    return userHasRoles(["admin"], user);
  })
```

**Example controller**

```js
  /**
   * Liste des posts pour l'admin
   */
  public async index({ view, request, bouncer }: HttpContextContract) {
    await bouncer.authorize("viewUsersList");
    const page = request.input("page", 1);
    const limit = 20;
    const entities = await Database.from(this.entityTable).paginate(
      page,
      limit
    );
    entities.baseUrl(this.entityListPath);
    return view.render(this.entityIndexView, { entities });
  }
```

**Comment by GabrielNBDS** [comment]
I simply add a `role` field to the user and create an ACL middleware.

**Model/User**
```ts
export default class User extends BaseModel {
  @column({ isPrimary: true })
  public id: number

  @column()
  public email: string

  @column()
  public name: string

  @column({ serializeAs: null })
  public password: string

  @column()
  public role: string
```

**Middleware/Acl.ts**
```ts
import { HttpContextContract } from '@ioc:Adonis/Core/HttpContext'

export default class Acl {
  public async handle(
    { auth, response }: HttpContextContract,
    next: () => Promise<void>,
    allowedRoles: string
  ) {
    const user = await auth.authenticate()

    if (!allowedRoles.includes(user.role)) {
      return response.unauthorized()
    }

    await next()
  }
}
```

**Adding Middleware to the route**
```ts
Route.resource('/anything', 'AnythingController')
  .apiOnly()
  .middleware({
    store: ['acl:admin,editor'],
    destroy: ['acl:admin'],
  })
```

**Reply by yann-yinn** [reply]
Thanks for sharing. Simple and effective.

How do you deal with hiding / showing part of the UI in your edge templates ? With bouncers, i can write:

```html
@can('adminEditPost', entity)
  <a href="{{entity._editLink}}">Editer</a> </td>
@end
```

**Reply by GabrielNBDS** [reply]
Cant you select the role from the user and make an condition based on that? I'm saying it because I don't use edge.

**Reply by yann-yinn** [reply]
I wrote how i was doing in my second post (defining bouncers, which allows use of "@can" in the templates), I was just wondering how you deal with that with your method :)

**Reply by yann-yinn** [reply]
You're right, we can probably simply access "auth.user.role" from the edge templates with your method, and create conditions on that.
